File: src/storeFx/unstar.ts
import { Fx } from '../_core/fx';
import { subsonicUnstar } from '../api/subsonic/methods/unstar';
import type { AppStore } from '../store/create';
import { fetchOneAlbum } from './fetchOneAlbum';
import { fetchOneArtist } from './fetchOneArtist';
import { fetchOneSong } from './fetchOneSong';

export const unstar = Fx.async(async function* f(params: {
  albumId?: string;
  artistId?: string;
  id?: string;
}) {
  const { store } = yield* Fx.ask<{ store: AppStore }>();
  const { credentials } = store.getState().auth;
  yield* subsonicUnstar(params).provide({ credentials });

  return params.albumId != null
    ? fetchOneAlbum(params.albumId)
    : params.artistId != null
      ? fetchOneArtist(params.artistId)
      : params.id != null
        ? fetchOneSong(params.id)
        : Fx.Ok();
});

File: src/storeFx/Unstar.ml
open Fx
open Monad_syntax

let make params =
  let* Deps.{ store } = ask () in
  let state = store |. Zustand.getState in
  let* () =
    Subsonic.unstar params |. provide { credentials = state.auth.credentials }
  in
  match params with
  | { albumId = None; artistId = None; id = None } -> ok ()
  | { albumId = Some albumId; _ } -> FetchOneAlbum.make albumId
  | { artistId = Some artistId; _ } -> FetchOneArtist.make artistId
  | { id = Some id; _ } -> FetchOneSong.make id
